File: 4 Estrategias Evolutivas/Códigos/2 (mu+1)-ES/ES.py
import matplotlib.pyplot as plt
import numpy as np
from Plot_Contour import plot_contour
from Plot_Surf import plot_surf

def Recombination(x1, x2):
    y = 0.5 * (x1 + x2)
    return y

f = lambda x, y: x * np.exp(-x**2 - y**2)
xl = np.array([-2, -2])
xu = np.array([2, 2])

# f = lambda x, y: (x-2)**2 + (y-2)**2
# xl = np.array([-5, -5])
# xu = np.array([5, 5])

G = 100
D = 2
mu = 30

x = np.zeros((D, mu+1))
sigma = np.zeros((D, mu+1))
fitness = np.zeros(mu+1)

for i in range(mu):
    x[:, i] = xl + (xu - xl) * np.random.rand(D)
    sigma[:, i] = 0.2 * np.random.rand(D)
    fitness[i] = f(x[0, i], x[1, i])

for g in range(G):
    plot_contour(f, x, xl, xu)

    r1 = np.random.randint(mu)
    r2 = r1

    while r2 == r1:
        r2 = np.random.randint(mu)

    x[:, mu] = Recombination(x[:, r1], x[:, r2])
    sigma[:, mu] = Recombination(sigma[:, r1], sigma[:, r2])

    r = np.random.normal(0, sigma[:, mu], D)
    x[:, mu] = x[:, mu] + r

    fitness[mu] = f(x[0, mu], x[1, mu])

    Idx = np.argsort(fitness)
    x = x[:, Idx]
    sigma = sigma[:, Idx]
    fitness = fitness[Idx]

Idx_b = np.argmin(fitness)
plot_surf(f, x, xl, xu)
print("Mínimo en x=", x[0, Idx_b], " y=", x[1, Idx_b], " f(x,y)=", f(x[0, Idx_b], x[1, Idx_b]))

File: 4 Estrategias Evolutivas/Códigos/2 (mu+1)-ES/ES.m
clear all
close all
clc

f = @(x,y) x.*exp(-x.^2-y.^2);
xl = [-2 -2]';
xu = [2 2]';

% f = @(x,y) (x-2).^2+(y-2).^2;
% xl = [-5 -5]';
% xu = [5 5]';

G = 300;
mu = 30;
D = 2;

x = zeros(D,mu+1);
sigma = zeros(D,mu+1);
fitness = zeros(1,mu+1);

for i=1:mu
    x(:,i) = xl+(xu-xl).*rand(D,1);
    sigma(:,i) = 0.2*rand(D,1);
    fitness(i) = f(x(1,i),x(2,i));
end

for t=1:G
    Plot_Contour(f,x,xl,xu); % Gráfica
    
    r1 = randi([1 mu]);
    r2 = r1;
    while r1==r2
        r2 = randi([1 mu]);
    end
    
    x(:,mu+1) = Recombination(x(:,r1),x(:,r2));
    sigma(:,mu+1) = Recombination(sigma(:,r1),sigma(:,r2));

    r = normrnd(0,sigma(:,mu+1),[D 1]);
    x(:,mu+1) = x(:,mu+1) + r;
    
    fitness(mu+1) = f(x(1,mu+1),x(2,mu+1));

    [~,I] = sort(fitness);
    x = x(:,I);
    sigma = sigma(:,I);
    fitness = fitness(I);
end

[~,ib] = min(fitness);

figure
Plot_Surf(f,x,xl,xu) % Gráfica
disp([' mínimo global en: x=' num2str(x(1,ib)) ', y=' num2str(x(2,ib)) ', f(x,y)=' num2str(f(x(1,ib),x(2,ib)))])

%% Funciones
function y = Recombination (x1,x2)
    y = 0.5*(x1+x2);
end
